pennylane/devices/qubit_mixed/apply_operation.py

```diff
@@ -672,3 +687,174 @@ def apply_snapshot(
+      2. Take the tensor product of the complement density matrix and the provided density_matrix.
+      3. Reshape to the correct final shape and return.
+
+    Args:
+        op (qml.QubitDensityMatrix): The QubitDensityMatrix operation.
+        state (array-like): The current quantum state.
+        is_state_batched (bool): Whether the state is batched.
+        debugger: A debugger instance for diagnostics.
+        **execution_kwargs: Additional kwargs.
+
+    Returns:
+        array-like: The updated quantum state.
+
+    Raises:
+        ValueError: If the density matrix is invalid.
+    """
+    density_matrix = op.parameters[0]
+    num_wires = len(op.wires)
+    expected_dim = 2**num_wires
+
+    # Cast density_matrix to the same type and device as state
+    density_matrix = math.cast_like(density_matrix, state)
+
+    # Validate shape
+    if math.shape(density_matrix) != (expected_dim, expected_dim):
+        raise ValueError(
+            f"Density matrix must have shape {(expected_dim, expected_dim)}, "
+            f"but got {math.shape(density_matrix)}."
+        )
+
+    # Validate Hermiticity
+    if not math.allclose(density_matrix, math.conjugate(math.transpose(density_matrix))):
+        raise ValueError("Density matrix must be Hermitian.")
+
+    # Validate trace
+    one = math.asarray(1.0 + 0.0j, like=density_matrix)
+    if not math.allclose(math.trace(density_matrix), one):
+        raise ValueError("Density matrix must have a trace of 1.")
+
+    # Extract total wires
+    num_state_wires = _get_num_wires(state, is_state_batched)
+    all_wires = list(range(num_state_wires))
+    op_wires = op.wires
+    complement_wires = [w for w in all_wires if w not in op_wires]
+
+    # If the operation covers the full system, just return it
+    if len(op_wires) == num_state_wires:
+        # If batched, broadcast
+        if is_state_batched:
+            batch_size = math.shape(state)[0]
+            density_matrix = math.broadcast_to(
+                density_matrix, (batch_size,) + math.shape(density_matrix)
+            )
+
+        # Reshape to match final shape of state
+        return math.reshape(density_matrix, math.shape(state))
+
+    # Partial system update:
+    # 1. Partial trace out op_wires from state
+    # partial_trace reduces the dimension to only the complement wires
+    sigma = qml.math.partial_trace(state, indices=op_wires)
+    # sigma now has shape:
+    # (batch_size?, 2^(n - num_wires), 2^(n - num_wires)) where n = total wires
+
+    # 2. Take kron(sigma, density_matrix)
+    sigma_dim = 2 ** len(complement_wires)  # dimension of complement subsystem
+    dm_dim = expected_dim  # dimension of the replaced subsystem
+    if is_state_batched:
+        batch_size = math.shape(sigma)[0]
+        sigma_2d = math.reshape(sigma, (batch_size, sigma_dim, sigma_dim))
+        dm_2d = math.reshape(density_matrix, (dm_dim, dm_dim))
```

Comment:
Unless @codecov is wrong, looks like these lines still miss test coverage (same below)